plugin/pulpcore/plugin/download/asyncio/factory.py

```diff
@@ -0,0 +1,116 @@
+import atexit
+from gettext import gettext as _
+import ssl
+from urllib.parse import urlparse
+
+from .http import HttpDownloader
+
+
+class DownloadFactory:
+    """
+    A factory for creating configured downloader objects that all share a single session.
+
+    Usage:
+        >>> import asyncio
+        >>> loop = asyncio._get_running_loop()
+        >>> the_factory = DownloadFactory(importer)
+        >>> downloader_a = the_factory.build(url_a)
+        >>> downloader_b = the_factory.build(url_b)
+        >>> downloads_list = [downloader_a, downloader_b]
+        >>> done, not_done = loop.run_until_complete(asyncio.wait(downloads_list))
+        >>> for task in done:
+        >>>     result = task.result()  # 'result' is a DownloadResult
+    """
+
+    def __init__(self, importer, download_class=HttpDownloader):
+        """
+
+        Args:
+            importer (pulpcore.plugin.models.Importer): The importer used to populate downloader
+                settings.
+            download_class (pulpcore.plugin.download.HttpDownloader): The asyncio-aware downloader
+                class this factory will use when instantiating downloaders. This option allows the
+                factory to work with custom downloaders. This is an optional parameter and defaults
+                to the HttpDownloader.
+        """
+        self._importer = importer
+        self._download_class = download_class
+        self._session = self._make_session_from_importer()
+        self._protocol_map = {'https': self._http_or_https, 'http': self._http_or_https}
+        atexit.register(self._session.close)
+
+    def _make_session_from_importer(self):
+        tcp_conn_opts = {}
+
+        sslcontext = None
+        if self._importer.ssl_ca_certificate.name:
+            sslcontext = ssl.create_default_context(cafile=self._importer.ssl_ca_certificate.name)
+            if self._importer.ssl_client_key.name and self._importer.ssl_client_certificate.name:
+                sslcontext.load_cert_chain(
+                    self._importer.ssl_client_key.name,
+                    self._importer.ssl_client_certificate.name
+                )
+        else:
+            if self._importer.ssl_client_key.name and self._importer.ssl_client_certificate.name:
+                sslcontext = ssl.create_default_context()
+                sslcontext.load_cert_chain(
+                    self._importer.ssl_client_key.name,
+                    self._importer.ssl_client_certificate.name
+                )
+
+        if sslcontext:
+            tcp_conn_opts['ssl_context'] = sslcontext
+
+        if self._importer.ssl_validation:
+            tcp_conn_opts['verify_ssl'] = self._importer.ssl_validation
+
+        conn = aiohttp.TCPConnector(**tcp_conn_opts)
+
+        auth_options = {}
+        if self._importer.basic_auth_user and self._importer.basic_auth_password:
+            auth_options['auth'] = aiohttp.BasicAuth(
+                username=self._importer.basic_auth_user,
```

Comment:
s/username/login
`class BasicAuth(namedtuple('BasicAuth', ['login', 'password', 'encoding'])):`

Reply:
Thanks. Fixed in next push.